WIP: MarshalWire

jaekwon opened the pull request:
MarshalWire for easy custom encoding;
Nil interfaces are encoded as 0x0000 (2 zeros) not 0x00000000 (4 zeros).
Nil interfaces are encoded as nil (0x01) in nillable lists;
Empty slices/arrays are not nil;

jaekwon committed:
First commit for MarshalWire;  Nil interfaces are encoded as nil in nillable lists;  Empty slices/arrays are not nil

jaekwon committed:
Fixing tests...

jaekwon committed:
Fix tests...

odeke-em commented on README.md:
```diff
@@ -334,14 +337,19 @@ elements of a list are encoded without an index or key.  They are just encoded
+To declare that the List may contain nil elements (e.g. the list "nillable"),
+the Lists's element typ4 byte should set the 4th least-significant bit (the
+"pointer bit") to 1.  If (and only if) the pointer bit is 1, each element is
+prefixed by a "nil byte" -- a 0x00 byte to declare that a non-nil item follows,
```
this seems a bit backwards:
* 0x00 to declare that a non-nil item follows
* 0x01 to declare that a nil item follows

You talk about why you flipped the values in the next sentence but how are we going to encode a uvarint in 1 byte?

jaekwon replied:
the nil byte precedes the encoding of the uvarint (or whatever are the elements of the list).

<element typ4 w/ pointer bit set to 1><uvarint(len(elements)=4)><0x01 (nil)><0x01 (nil)><0x00 (not nil)><encoding of 3rd element><0x01 (nil)> as an example with 4 elements.

odeke-em commented on binary-decode.go:
```diff
@@ -81,7 +94,25 @@ func (cdc *Codec) _decodeReflectBinary(bz []byte, info *TypeInfo, rv reflect.Val
-	var _n int
+	// Handle override if a pointer to rv implements UnmarshalWire.
+	if info.IsWireUnmarshaler {
+		// First, decode repr instance from bytes.
```
So why don't we just invoke UnmarshalWire directly and let the method handle all this?

The second call to _decodeReflectBinary on line 105 worries me about endless recursion.

jaekwon replied:
mmm then it would be less type-safe because MarshalWire/UnmarshalWire would have to specify the repr-type as only interface{} or equally generic.  As specified, due to the flexibility of the method signature, we cannot make this an interface... but it's not really a problem for either reflection or code generation.

odeke-em commented on binary-decode.go:
```diff
@@ -81,7 +94,25 @@ func (cdc *Codec) _decodeReflectBinary(bz []byte, info *TypeInfo, rv reflect.Val
+			return
+		}
+		// Then, decode from repr instance.
+		uwrm := rv.Addr().MethodByName("UnmarshalWire")
```
Nice!

odeke-em commented on wire.go:
```diff
@@ -15,7 +15,7 @@ import (
 
 type Typ3 uint8
-type Typ4 uint8 // Typ3 | 0x80 (pointer bit)
+type Typ4 uint8 // Typ3 | 0x08 (pointer bit)
```
This perhaps warrants us removing a bunch of these magic constants in favour of variables/consts
```go
const typ4Pointer uint8 = 0x08
```
and I like the new introduction of the method `IsPointer() bool`

jaekwon replied:
Yeah we should move over to methods everywhere.

odeke-em commented on wire.go:
```diff
@@ -108,15 +109,12 @@ func (cdc *Codec) MarshalBinary(o interface{}) ([]byte, error) {
-			panic("MarshalBinary cannot marshal a nil pointer.")
-		}
+	// Dereference value if pointer.
+	var rv, _, isNilPtr = derefPointers(reflect.ValueOf(o))
```
Aha, so it returns.

jaekwon replied:
Yeah it's basically the same logic.

jaekwon committed:
Add protobuf parsing image to start writing cmd/wirescan docs

jaekwon committed:
Update README

jaekwon committed:
Update README.md

jaekwon committed:
Update README.md

jaekwon committed:
Update README.md

jaekwon committed:
Update README.md

Chjango committed:
Minor typos and semantics edits

Removed our call for bounties at the very end and changed it to scouting for contributors instead since we're not paying out developer bounties.

Chjango committed:
Minor edits

Spelling and other error removal. Improved some wording.

jaekwon committed:
Update README.md

jaekwon committed:
Update README.md

jaekwon committed:
Merge pull request #2 from Chjango/patch-1

Minor typos and semantics edits

Commit:
README: comparison with JSON updates, Go vs Golang (#3)

* Comparison with JSON updates

* The language is Go not Golang; golang is the website and
the hashtag even though commonly called "Golang" ;)

* Comparison with Protobuf3 to state that embedded messages
are length prefixed in Protobuf3 while in wire they aren't

* Update on RegisterInterface and RegisterConcrete for clarity

martindale committed:
Improve formatting and readability (#1)

* Improve formatting and readability
* Re-nest `oneof` description in comparison
